text/subsingleton-elimination.md

```diff
@@ -0,0 +1,176 @@
+
+# Detailed design for the `Prop` and `SProp` extensions
+
+## Uniformizing the treatment between `Prop` and `SProp`
+
+1. accept singleton types with all arguments in `SProp` to be in `SProp` as it is the case for `Prop`, the only difference being that recursive arguments are excluded; the indices should be in small types (i.e. `Prop` or `Set`)
+
+## Generalize the syntactic criterion for being in `Prop` or `SProp`
+ 
+2. accept in `SProp` and `Prop` singleton types with arguments of the form `forall a:A, B` when `B` is recursively an `SProp` or `Prop`-subsingleton
+3. accept in `SProp` and `Prop` types with several constructors when (small) indices justify that the constructors are disjoint (section 5 of "Definitional Proof-Irrelevance without K")
+
+## Dynamically recognizing subsingleton elimination for `Prop` and `SProp` in `match`
+
+4. This would require calling the subsingleton elimination checker dynamically when typing `match`.
+
+# Detailed design for the `HProp` addition
+
+Belonging to `HProp` is not decidable, so exactly characterizing `HProp` requires user-provided data.
+
+There are two solutions:
+- either we retarget `Prop` to include `HProp` (simpler)
+- or we keep `Prop` as it is and add a new subuniverse `HProp` (less change of habits)
+
+Both solutions follow the same way. Below, we assume the latter. To get the former, just use `Prop` instead of `HProp`.
+
+We shall avoid relying on template polymorphism. However, we may consider a `cast` of the form:
+```
+(t : HProp by p)
+```
+which forces the type-checker to recognize a type as being in a smaller universe than decidably detectable.
+
+The status of casts in the implementation is however fragile, so this may require some clarification of casts first.
+
+## New command to support arbitrary `HProp`
+
+We propose a new command:
+```
+Subsingleton id context : ishProp A := proof.
+```
+which recast the type of `A` from `context -> Type` to `context -> HProp` so that it is latter considered impredicative.
+
+An alternative is to have a modifier to definitions in the form:
+```
+Subsingleton id context := def : HProp by proof.
+```
+But then, we probably want also:
+```
+Subsingleton id context := def : Hprop.
+Proof.
+the proof of being hProp
+Qed.
+```
+which may be complicated.
+
+If `HProp` is made distinct from `Prop`, new declarations `Inductive ... : ... -> HProp` would be accepted too. When to tell that the type has subsingleton elimination? There is a proposal in the next section.
+
+## Dynamic detection of subsingleton elimination
+
+We propose an extension of `match` of the following form:
+```
+match t as id in I params return typ by u with
+| C1 vars => body
+...
+| Cn vars => body
+end
+```
+where, when `I` is `HProp`-truncated but the instance `I args` is an `HProp` proved by `u : forall x y:I args, x = y`, the elimination to `Type` is allowed.
```

Comment:
If I is a truncation then `x = y :> I _` is always inhabited, ie we can't express "I may be un-truncated" in terms of I.

Reply:
Yes, my wording is wrong. The term `u` should be a proof of `forall x y:I args, x = y` where, in the proof, `I` is taken in `Type`, i.e. not truncated, even if in the `match`, it is truncated.

Maybe another criterion would be to prove that `forall vars vars', vars = vars'` but the equality is complicated to express when there are dependencies.

Reply:
Actually, you have `x = y :> I _` always inhabited for `SProp` but not for `Prop` because nothing enforces that `Prop` is actually proof-irrelevant (there is a model of `Prop` which interprets it as impredicative `Set` for instance).